Rework split, layout, workspace_layout, and move commands to work like i3

RPigott opened the pull request:
EDIT: Hijacking top comment b/c this PR is barely related to what it started as anyway

Hey you. Lurker. Please test.

---
Fix #5755 

This is intended to match i3's behavior here:

https://github.com/i3/i3/blob/3cd1c45eba6de073bc4300eebb4e1cc1a0c4479a/src/tree.c#L354

Xyene commented:
Wow, you beat me to the punch here -- I was going to make the same change tonight.

However, this isn't quite right (I know you haven't marked it as ready for review yet etc., but anyway). The check for the parent layout to be a [vh]split is actually necessary, because if you have a `T[app1]` and vsplit, this *should* result in `T[V[app1]]`, _not_ `V[app1]`.

Probably a good idea to leave a comment linking to the i3 source here, because this is pretty non-obvious.

RPigott commented:
Fixes #5755, #5754, Re-fixes #5157,  #4038.

I ended up going through a bunch of i3-compat issues but I think this matches i3 in the above cases. I don't think 5755 is related to 5754 just my fixes ended up on the same branch.

Yeah thanks @Xyene I forgot to copy that part. I added a comment as suggested.

RPigott commented:
For #4038 OP seems to describe i3 behavior as `T[1 2 3] -> H[T[1 2] T[3]]` but I can't reproduce this. In i3 I get `T[1 2 3] -> T[T[1 2] T[3]]`.

Xyene commented:
> In i3 I get `T[1 2 3]` -> `T[T[1 2] T[3]]`.

I'm not entirely sure how... `workspace_layout normal`, open two windows:

![image](https://user-images.githubusercontent.com/1403503/97056270-92d90900-1556-11eb-8b64-8047e6c82dac.png)

`layout tabbed`:

![image](https://user-images.githubusercontent.com/1403503/97056280-9cfb0780-1556-11eb-90ef-2244b853332c.png)

`move right`:

![image](https://user-images.githubusercontent.com/1403503/97056291-a4baac00-1556-11eb-9f59-988d09cc45c7.png)

This state is not correct. Subsequent `move right`s from here toggle between `T[app1 T[app2]]` and `T[app1 app2]`. Instead, the transition should be `T[app1 app2]` &rarr; `H[T[app1] app2]`

RPigott commented:
Yeah this is wrong in that case.

> This only differs in i3 in that i3 always has the wrapping H for the workspace, whereas Sway often changes the workspace layout itself.

i3 doesn't always wrap workspaces, just when changinging the layout on a single container. Otherwise they change the workspace layout. The difference here is in the behavior of the layout command, which I guess I should also update to match:

https://github.com/i3/i3/blob/3cd1c45eba6de073bc4300eebb4e1cc1a0c4479a/src/con.c#L1817

The case I was trying to get at (and how I interpreted #4038) is

Begin: empty workspace

`layout tabbed`, open apps, `move right`. Then the workspace layout changes and I get the `T[1 2 3] -> T[T[1 2] T[3]]` behavior.

Xyene commented:
Getting better :+1: 

The case I described still happens with `workspace_layout tabbed`, though it has been fixed for `workspace_layout normal`.

RPigott commented:
> i3 workspace_layout doesn't change the layout of the root workspace container; it changes that of its first-level children.

Wow no kidding. In i3 `workspace_layout tabbed` doesn't make tabbed workspaces but `layout tabbed` on the workspace does.

The sway(5) man page will probably need to be updated with a proper description of `workspace_layout` too.

Xyene commented:
Yeah, it's tricky. Bit of a meta-comment, and you may already have a testing setup that works for you, but I've found the following to be very helpful in comparing i3 with Sway:

* script to dump layout, works on both i3 and Sway
```
$ cat ~/gettree.sh                                                                                                                                    
swaymsg -t get_tree|jq "def f: {name, type, layout, nodes: [.nodes[]|f]}; recurse(.nodes[])|select(.name == \"$1\")|f"
$ ~/gettree.sh 1   # get tree of workspace 1
```

* start Xephyr

```
$ Xephyr +iglx -br -ac -resizeable -noreset -keybd ephyr,,xkbmodel=evdev :2 &
```

* start i3 within Xephyr

```
$ unset XDG_RUNTIME_DIR && unset WAYLAND_DISPLAY && DISPLAY=:2 i3
```

* passthrough mode in Sway config so you can use conflicting keybinds within Xephyr i3

```
mode "passthrough" {
	bindsym $mod+p mode default
}

bindsym $mod+p mode passthrough
```

RPigott commented:
So far I've been running i3 in a separate session and sway embedded in gdb. It's a workable setup but could be improved. The i3 layout dump will certainly be handy, I had a basic one but it was still bit of a hassle to interpret. I need to test more i3 layouts

Quick take at matching i3 workspace_layout behavior like https://github.com/i3/i3/blob/e54e88b9e58579b406c5de94b68cb9ad3211eb4e/src/workspace.c#L899, but I'm guessing code will need to be moved somewhere more appropriate than view_map to match all the cases.

Xyene commented:
The latest commit certainly improves things. I'm generally a fan of keeping the workspace node layout intact (i3-style) over what Sway currently does; it makes debugging i3-compat issues easier.

With `workspace_layout tabbed`, starting with this layout and `move right`:

<details><summary>Start state</summary>

```json
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        },
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    }
  ]
}
```
</details>

<details><summary>Sway result</summary>

```json
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": "xyene@lynwood: /code/sway-src/sway",
      "type": "con",
      "layout": "none",
      "nodes": []
    }
  ]
}
```
</details>

<details><summary>i3 result</summary>

```json
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/i3",
          "type": "con",
          "layout": "splith",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/i3",
          "type": "con",
          "layout": "splith",
          "nodes": []
        }
      ]
    }
  ]
}
```
</details>

RPigott commented:
Hopefully improved. I Only tested it in a few cases and had #5761 applied on top.

I moved the change into workspace_add_tiling where it obviously belongs, but lots of places aren't happy to have workspace_add_tiling put their con as a child of a new container instead of the workspace they requested. So, most calls of workspace_add_tiling are changed to something like `con = workspace_add_tiling(ws, con)` and some of the following references are changed to modify the new con, if there is one.

Xyene commented:
That case works much better now. Quick testing: starting with a `T[app]` and doing repeated `splitv`, you end up with `V[V[V[...[V[app]]]]]`.

Xyene commented:
One more weird case with `workspace_layout tabbed` where a blank container was created:

<details><summary>Log</summary>

```
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1                                                                                                                                                                                                                                                        
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1                                                                                                                                                                                                                                                        
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": "xyene@lynwood: /code/sway-src/sway",
      "type": "con",
      "layout": "none",
      "nodes": []
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1
{                                                                                                                                                          
   "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null, <--------------------------------- this container is blank
          "type": "con",
          "layout": "splith",
          "nodes": []
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up                                                                                    
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": "xyene@lynwood: /code/sway-src/sway",
      "type": "con",
      "layout": "none",
      "nodes": []
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": []
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up                                                                                    
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": []
        }
      ]
    }
  ]
}
```

</details>

RPigott commented:
> Quick testing: starting with a `T[app]` and doing repeated `splitv`, you end up with `V[V[V[...`

Oops, pushed the corrected version of that commit.

> One more weird case with workspace_layout tabbed where a blank container was created

I don't observe that with your #5761 applied. I guess it gets reaped.

Xyene commented:
Exciting, it's getting there :)

> I don't observe that with your #5761 applied. I guess it gets reaped.

You're right, oops.

It does seem that some extraneous containers are still being created with `workspace_layout tabbed`:

<details><summary>Log</summary>

```
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1                                                                                        
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        },
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move right
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1
{                                                                          
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1   
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": [
            {
              "name": null,
              "type": "con",
              "layout": "tabbed",
              "nodes": [
                {
                  "name": "xyene@lynwood: /code/sway-src/sway",
                  "type": "con",
                  "layout": "none",
                  "nodes": []
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up    
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1    
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": [
            {
              "name": null,
              "type": "con",
              "layout": "tabbed",
              "nodes": [
                {
                  "name": null,
                  "type": "con",
                  "layout": "splith",
                  "nodes": [
                    {
                      "name": null,
                      "type": "con",
                      "layout": "tabbed",
                      "nodes": [
                        {
                          "name": "xyene@lynwood: /code/sway-src/sway",
                          "type": "con",
                          "layout": "none",
                          "nodes": []
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ swaymsg move up    
(RPigott-consplit) xyene@lynwood:/code/sway-src/sway$ ~/gettree.sh 1    
{
  "name": "1",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": "xyene@lynwood: /code/sway-src/sway",
          "type": "con",
          "layout": "none",
          "nodes": []
        }
      ]
    },
    {
      "name": null,
      "type": "con",
      "layout": "tabbed",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": [
            {
              "name": null,
              "type": "con",
              "layout": "tabbed",
              "nodes": [
                {
                  "name": null,
                  "type": "con",
                  "layout": "splith",
                  "nodes": [
                    {
                      "name": null,
                      "type": "con",
                      "layout": "tabbed",
                      "nodes": [
                        {
                          "name": null,
                          "type": "con",
                          "layout": "splith",
                          "nodes": [
                            {
                              "name": null,
                              "type": "con",
                              "layout": "tabbed",
                              "nodes": [
                                {
                                  "name": "xyene@lynwood: /code/sway-src/sway",
                                  "type": "con",
                                  "layout": "none",
                                  "nodes": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

</details>

RPigott commented:
The commit I removed was just one that had commented out some code. Now workspace_wrap_children now doesn't double wrap, that was creating some extra containers.

So here's an interesting layout with `workspace_layout tabbed`:

`begin` -> `H[T[app app*]]`
`move right` -> `H[T[app] T[app*]]` (matches i3)
`move up` -> `V[T[app*] H[T[app]]]` (matches i3, your scenario above)
`move right` -> `H[V[H[T[app]]] T[app*]]` (sway), `H[T[app] T[app*]]` (i3)

In the last move, though it's pretty unwieldy, the sway layout is the one that matches my mental model of the i3 tree. I'd guess it's special cased somewhere, but atm I'm not sure what i3 is doing here to get the simpler layout.

Xyene commented:
I'd guess it's this:

https://github.com/i3/i3/blob/1f0c628cde40cf87371481041b7197344e0417c6/src/tree.c#L638-L651

```c
/*
 * tree_flatten() removes pairs of redundant split containers, e.g.:
 *       [workspace, horizontal]
 *   [v-split]           [child3]
 *   [h-split]
 * [child1] [child2]
 * In this example, the v-split and h-split container are redundant.
 * Such a situation can be created by moving containers in a direction which is
 * not the orientation of their parent container. i3 needs to create a new
 * split container then and if you move containers this way multiple times,
 * redundant chains of split-containers can be the result.
 *
 */
void tree_flatten(Con *con) {
```

Called right at the end of `tree_move` on `croot`, which looks like, well, the root of the workspace: https://github.com/i3/i3/blob/1f0c628cde40cf87371481041b7197344e0417c6/src/move.c#L388

RPigott commented:
Hmm, thanks for the tip.

I implemented "workspace_squash" which is basically equivalent to i3's tree_flatten. I called it squash since container flatten was already taken to go the other way. Now I squash the workspace after moving and I get the same layout as i3 :tada:.

Xyene commented on sway/tree/workspace.c:
```diff
@@ -817,3 +833,10 @@ size_t workspace_num_tiling_views(struct sway_workspace *ws) {
+
+void workspace_squash(struct sway_workspace *workspace) {
+	for (int i = 0; i < workspace->tiling->length; i++) {
+		struct sway_container *child = workspace->tiling->items[i];
```
Can this access out of bounds memory if `workspace->tiling` shrinks in length during `container_squash`?

Xyene replied:
Separately, I'm not really sure whether we should be relying on the order containers appear in `workspace->tiling` here (for the skipping). It seems fragile if we don't already do this elsewhere.

RPigott replied:
I don't think workspace->tiling can shrink in length during the squash. container_squash only modifies the parent by adding containers to it, which is why I keep track of the number added. Containers added to the parent are not themselves eligible to be squashed because the recursion ends when it's done adding them and destroys the container pair. Basically each child has the opportunity to squash one pair of containers in its family, but it stops there.

Regardless, unless I'm sorely mistaken about C (possible), the condition statement is evaluated on each loop, so if length changes the condition value will reflect that and the loop will end rather then set `i >= workspace->length->tiling`. I expect that the length will grow because I am adding containers to it, so I actually need that for correctness. Otherwise the loop will end early.

Actually there is a bug though...  the reap_empty call should decrement change before returning because at that point con and child are certainly empty and will be removed. Atm I guess it skips the immediate sibling of children it successfully squashed.

Xyene replied:
You're right; I wasn't paying close enough attention.

Xyene commented on sway/commands/move.c:
```diff
@@ -366,21 +330,23 @@ static bool container_move_in_direction(struct sway_container *container,
-
-		current = current->parent;
+		if (!current->parent && !escaped) {
+			// No parallel parent, container is trying to escape it's workspace
```
```suggestion
			// No parallel parent, container is trying to escape its workspace
```

Xyene reviewed (commented):
Awesome! I can give this more thorough review / testing tonight. Could you rebase on master? A few conflicts have accumulated.

RPigott commented:
Rebased. I tried to add some more helpful commit messages since they were previously not too helpful.

I absorbed your commit because I think it should apply before the workspace_layout changes so that the intermediate commits  don't produce weird null containers. Hope that's alright.

I'll continue testing it for a while.

RPigott commented on include/sway/tree/workspace.h:
```diff
@@ -110,14 +110,14 @@ void workspace_unwrap_children(struct sway_workspace *ws,
-void workspace_insert_tiling(struct sway_workspace *workspace,
-		struct sway_container *con, int index);
+struct sway_container *workspace_insert_tiling(struct sway_workspace *workspace,
+		struct sway_container *con, int index, enum sway_container_layout layout);
```
I added this layout argument because I needed a fucntion that performed the insert but without the potential split for workspace_layout.

Since this ended up only being used in one place with a layout other than config->default_layout I'd rather remove it and make that a different function.

Xyene replied:
Yeah, that sounds like a good idea.

Xyene replied:
Yeah, that sounds like a good idea.

Xyene commented:
> I absorbed your commit because I think it should apply before the workspace_layout changes so that the intermediate commits don't produce weird null containers. Hope that's alright.

That's great, I've closed the other PR.

One immediate thing I've noticed is that I can no longer move containers across workspaces when one workspace is on another monitor (with `move left/right` when at the edge of a monitor).

Xyene commented:
I believe the "resizing based off siblings" logic regressed a bit, namely in this layout:

<details><summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/1403503/97925523-88c1c200-1d2f-11eb-9fa5-598318b5acc9.png)

`move down`

![image](https://user-images.githubusercontent.com/1403503/97925554-99723800-1d2f-11eb-85dc-34f1ef27ef4a.
</details>

in i3, the resulting layout is:

<details><summary>i3 layout</summary>

![image](https://user-images.githubusercontent.com/1403503/97925692-d63e2f00-1d2f-11eb-87b3-bae00ecd27e7.png)

</details>

Xyene commented on include/sway/tree/workspace.h:
```diff
@@ -134,4 +134,6 @@ size_t workspace_num_tiling_views(struct sway_workspace *ws);
 
 size_t workspace_num_sticky_containers(struct sway_workspace *ws);
 
+void workspace_squash(struct sway_workspace *workspace);
```
Let's add a description of both what this and `container_flatten` does to the headers.

Xyene commented on sway/tree/container.c:
```diff
@@ -797,16 +797,16 @@ void container_set_floating(struct sway_container *container, bool enable) {
-			workspace_add_tiling(workspace, container);
-			container->width = workspace->width;
-			container->height = workspace->height;
+			other = workspace_add_tiling(workspace, container);
```
Let's just declare a new variable for this here rather than taking over `reference`/`other` from above to mean a different thing depending on what branch you're in.

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1619,68 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
 	return con->scratchpad && !con->workspace;
 }
+
+static enum sway_container_layout layout_category(enum sway_container_layout layout) {
```
Nit: `container_...` naming scheme to match the rest of this file.

RPigott replied:
I replaced this helper with a new helper function "is_parallel" similar to commands/move

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1619,68 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
+int container_squash(struct sway_container *con) {
+	int change = 0;
+	if (!con->children) {
+		return change;
```
Always 0?

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1619,68 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
+	}
+	if (con->children->length != 1) {
+		container_squash_children(con);
+		return change;
```
Looks like this is also always 0; let's make that explicit here and move the `int change = 0;` declaration to L1668.

RPigott replied:
Seems good. Technically I could just declare it as `int change = -1` instead of ending with `change--` since I know in advance I will reap con. I'll just keep it as 0 though I think. I prefer that the value matches its meaning throughout.

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1619,68 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
+		change--;
+		return change;
+	}
+	container_squash_children(con);
```
The `if` above ends with a `return`, so we can put this line in an `else` block, and deduplicate the return by removing the one on L1682 and keeping the one on L1685.

Xyene commented on sway/commands/layout.c:
```diff
@@ -162,6 +162,13 @@ struct cmd_results *cmd_layout(int argc, char **argv) {
+			container = workspace_wrap_children(workspace);
+			container->layout = new_layout;
+			container_update_representation(container);
+			workspace_update_representation(workspace);
```
I think this call is implied in `container_update_representation`, is it not? https://github.com/swaywm/sway/blob/39328ca4e4b97bd76153869f6a0a20fef3c17112/sway/tree/container.c#L612-L616

RPigott commented:
> I believe the "resizing based off siblings" logic regressed a bit, namely in this layout:

Yes I noticed that too. I'll look into it.

> One immediate thing I've noticed is that I can no longer move containers across workspaces when one workspace is on another monitor (with move left/right when at the edge of a monitor).

Still works for me when the container is at workspace level, but obviously that's not as common anymore with all the layout changes. I'll check how i3 does it.

RPigott commented:
Aight, I'm back.

I had some trouble getting the move behavior to match i3 within the existing control flow, so in the end I basically just looked at i3's tree_move and translated it to sway. So a lot of the `container_move_in_direction` function has changed, but it seems to work.

For the geometry I forgot to reset the container height/width after wrapping a container – that should be better now.

For the moving between outputs i3 does special case those lone children of workspace level splits, so I added that as well in commands/move.c:371.

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1620,79 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
+		i += container_squash(child);
+	}
+}
+/*
```
I think this description is good, but ought to be moved to the header file since that's where most of the docs are.

Xyene commented on sway/tree/workspace.c:
```diff
@@ -830,3 +851,14 @@ size_t workspace_num_sticky_containers(struct sway_workspace *ws) {
+// workspace_squash is container_flatten in the reverse
+// direction. Instead of eliminating redundant splits that are
+// parents of the target container, it eliminates pairs of
+// redundant H/V splits that are children of the workspace.
```
Same comment re: header here.

Xyene reviewed (commented):
Thanks again for working on this -- it's getting better and better. I've left a couple of code style nits, and have started running on this branch. I haven't run into _major_ issues _yet_, so that's a good sign :)

Here's at least one small one, though: if you have a display layout like `| output 1 | output 2 |` with `T[H[app1*] app2]` on output 2, `move left` should result in `T[app1 app2]`. Instead, however, `app1` escapes onto output 1, as if the original layout had been `T[app1 app2]`.

Xyene commented:
Actually, one more thing: if you have a layout like `S[S[S[S[H[app]]]]]` and `move left`, only one level of the `S[` should be collapsed. Currently it collapses the entire chain.

RPigott commented:
Hmm yeah. I may be "squashing" too liberally. i3's move can return early and skip the squash sometimes, so I may need to move the squash call into specific branches of the move function.

RPigott commented:
> Actually, one more thing: if you have a layout like S[S[S[S[H[app]]]]] and move left, only one level of the S[ should be collapsed.

Wait are you sure? I just tried this in i3 and it collapsed all the S for me.

Still, I'll have to correct it incorrectly flattening the new ws if the con switches workspaces or unnecessarily flattening parents for simple moves.

Xyene commented:
> Wait are you sure? I just tried this in i3 and it collapsed all the S for me.

Uhhh... I _was_, but clearly I'm wrong. However, I did just confirm that if you had `H[V[H[V[app]]]]` and did `move left`, the result in i3 is `H[V[app]]`, but in this branch is just `H[app]`. This does seem like a case of over-squashing, as you say. (Also, I think this is not all running in the same transaction, or something: I saw flashes of `H[]` titlebars when moving the last child of a container out.)

The geometry issues also seem mostly fixed, though I was able to generate a "bad" geometry:

<details><summary>Screenshots</summary>

initial

![image](https://user-images.githubusercontent.com/1403503/98492154-f163df80-2204-11eb-8bcf-f70bd1a62bb8.png)

`move down`

![image](https://user-images.githubusercontent.com/1403503/98492165-fcb70b00-2204-11eb-81fb-31931aad90a4.png)

`move up`

![image](https://user-images.githubusercontent.com/1403503/98492173-0476af80-2205-11eb-8d10-4ba9a91e9aac.png)

`move down` again -- space ratio is wrong

![image](https://user-images.githubusercontent.com/1403503/98492195-15272580-2205-11eb-9a29-5edcf05a0d61.png)



</details>

RPigott commented:
> I did just confirm that if you had `H[V[H[V[app]]]]` and did move left, the result in i3 is `H[V[app]]`, but in this branch is just `H[app]`.

I just tried this in i3 and got `H[app]` as well...

<details>
<summary>
layouts
</summary>

```
$ gettree.sh 2 # HVHV[
{
  "name": "2",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "splitv",
      "nodes": [
        {
          "name": null,
          "type": "con",
          "layout": "splith",
          "nodes": [
            {
              "name": null,
              "type": "con",
              "layout": "splitv",
              "nodes": [
                {
                  "name": null,
                  "type": "con",
                  "layout": "splith",
                  "nodes": []
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
$ i3-msg move left
[{"success":true}]
$ gettree.sh 2 # H[
{
  "name": "2",
  "type": "workspace",
  "layout": "splith",
  "nodes": [
    {
      "name": null,
      "type": "con",
      "layout": "splith",
      "nodes": []
    }
  ]
} 
```
</details>

Could it be a config difference?

> `H[T[V[app1*]] T[app2]]` + `move right` is a no-op, though the result should be `H[T[app1] T[app2]]` [workspace_layout=tabbed]

Oops, accidentally caught by the wrong branch. I checked for workspace when I meant !parent – that probably messed up a lot of similar layouts.

I think this branch is getting much closer to a working state. I'll try to push smaller fixup commits and squash/rebase/whatever them later rather than re-writing half the branch each time. Hopefully that makes it easier to see what's changed.

Changing when it squashes doesn't fix any previously mentioned bugs I don't think, but should correct cases where you perform a simple move like a sibling swap within a deeply nested squashable layout. i3 leaves those layouts be if it didn't do anything that could introduce new splits during the move.

RPigott commented:
If it helps, my understanding of the `H[V[H[V[` case step-by-step is

`H[V[H[V[app]]]]` -`move left`-> `H[V[H[app V[]]]` -squash/flatten-> `H[app V[]]` -reap-> `H[app]`

You could get `H[V[app]]` with

`H[V[H[V[app]]]]` -`move up`-> `H[V[app H[V[]]]` -reap-> `H[V[app]]`

That's what I observe in both this branch and i3.

RPigott commented on sway/commands/move.c:
```diff
@@ -316,72 +306,88 @@ static bool container_move_in_direction(struct sway_container *container,
-		sway_log(SWAY_DEBUG, "Hit edge of output, nowhere else to go");
 	}
+	// Unreachable
 	return false;
```
So I added that there because the compiler gets upset that it reaches the end of the function without a return, and it does that because it doesn't know that since I changed the loop, "current" will actually never be null.

So the real problem here is I silenced the compiler warning without correcting the code. The loop condition should actually just be `while(true)` – the behavior is the same but it doesn't mislead readers, and the compiler, to think there is an invariant there.

Instead I think I'll move the declarations from the middle of the loop to before the loop, and the if block at the end to outside the loop. Once it gets to that final block it can't repeat anyway since every branch of the if has a return.

RPigott replied:
Actually better yet I can just introduce a variable to be the loop invariant.

Xyene commented:
> I just tried this in i3 and got H[app] as well...

Hmm, I think you're probably right and I wasn't running the same command, but at any rate I can no longer generate weird behavior after your fixups.

> Oops, accidentally caught by the wrong branch. I checked for workspace when I meant !parent – that probably messed up a lot of similar layouts.

> I think this branch is getting much closer to a working state.

Agreed, good job :)

> I'll try to push smaller fixup commits and squash/rebase/whatever them later rather than re-writing half the branch each time. Hopefully that makes it easier to see what's changed.

This sounds fine, I was using the link under "force-pushed" to view the changes, but this sounds easier on your end for now.

At the moment, the only remaining issue I've seen is the geometry one from https://github.com/swaywm/sway/pull/5756#issuecomment-723710972. I'll continue running on this branch for a while (and with `workspace_layout tabbed`), just in case new things pop up.

RPigott commented:
> This sounds fine, I was using the link under "force-pushed" to view the changes, but this sounds easier on your end for now.

Wow I didn't know about that feature. Anyway, thanks a million for reviewing and testing stuff.

> At the moment, the only remaining issue I've seen are:
> - the geometry one from #5756 (comment), and
> - the transaction one from the same comment

The geometry one at least is an existing issue in sway, not introduced by this PR.

It's not super easy to get the H[T[V[ layout from your example on master but if you do set it up and try moving the second app down/up/down (it takes a few more movements than it does on this branch b/c of the changes) you can get the same stunted lower window.

I'm not sure about the issue with container titles, it may have been introduced by b149e46 since it now squashes before reaping any empty containers. Maybe reaping should be part of squash's duties.

RPigott commented:
I don't know why it included the wlroots 0.12 commit, that was a mistake. The only change is the new commit intended to address the geometry of promoted containers. I squashed the fixup commits from before.

Xyene commented on sway/commands/move.c:
```diff
@@ -263,138 +265,141 @@ static void container_move_to_container(struct sway_container *container,
+					move_dir == WLR_DIRECTION_LEFT ||
+					move_dir == WLR_DIRECTION_RIGHT ?
+					L_HORIZ : L_VERT;
+				container->height = container->width = 0;
```
Somewhat an aside from the meat of this PR, but I think we should have something like a `container_reset_geometry` for this since these lines appear in a lot of other places.

RPigott replied:
You're probably right, but I think I'll hold off adding that in this PR.

Xyene commented on sway/tree/container.c:
```diff
@@ -1609,3 +1620,69 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
+		struct sway_container *child) {
+	enum sway_container_layout gp_layout = con->parent ?
+		con->parent->layout : con->workspace->layout;
+	enum sway_container_layout con_layout = con->layout;
```
Hmm, is there a reason why these are stored in variables here? `is_parallel` is pure, and the logic below mixes `con->layout` with `con_layout`, and `child->layout` with `child_layout`.

Xyene commented on include/sway/tree/workspace.h:
```diff
@@ -110,13 +110,16 @@ void workspace_unwrap_children(struct sway_workspace *ws,
 		struct sway_container *con);
 
-void workspace_insert_tiling(struct sway_workspace *workspace,
+void workspace_insert_tiling_direct(struct sway_workspace *workspace,
```
We should have a comment here explaining the difference in nuance between `workspace_insert_tiling_direct` and `workspace_insert_tiling`. I only understand it since I've been following along this thread :)

Xyene reviewed (commented):
> Wow I didn't know about that feature. 

It's very well-hidden :)

> Anyway, thanks a million for reviewing and testing stuff.

Thank _you_ for working on this -- I know it expanded way further than you probably imagined it would, but thanks for sticking through it.

> I'm not sure about the issue with container titles, it may have been introduced by b149e46 since it now squashes before reaping any empty containers. Maybe reaping should be part of squash's duties.

I _feel_ like some function the code is calling is calling `transaction_commit_dirty` when it shouldn't. Alternate explanation: some internal state (probably geometry!) is trashed prematurely, causing a transaction to time out. This is a snippet of the Sway log on executing a `move up` that resulted in a container title flash:

```
00:15:36.793 [sway/desktop/transaction.c:428] Transaction 0x55d3cbabb900 committing with 10 instructions
00:15:36.993 [sway/desktop/transaction.c:389] Transaction 0x55d3cbabb900 timed out (1 waiting)
00:15:36.993 [sway/desktop/transaction.c:281] Applying transaction 0x55d3cbabb900
00:15:36.994 [sway/desktop/transaction.c:428] Transaction 0x55d3cb99f880 committing with 8 instructions
00:15:36.994 [sway/desktop/transaction.c:281] Applying transaction 0x55d3cb99f880
```

That looks like 3 calls to `transaction_commit_dirty`, one of which times out after 200ms. That causes the title to show the incorrect result for 200ms. I think either not scheduling that failing transaction in the first place, or figuring out why it's hanging would be sufficient to address this issue.

> The only change is the new commit intended to address the geometry of promoted containers.

Awesome!

RPigott commented:
- Fixed a bug in the first commit where floating containers couldn't be split.
- Fixed a bug where parents of promoted containers could sometimes survive being emptied

Actually, @Xyene does that fixup commit fix the titles for you? I may have been getting lucky and those accidental containers were just getting reaped later.

Xyene commented:
> Actually, @Xyene does that fixup commit fix the titles for you? 

Yes, the titles are fine now and transactions are no longer timing out. Great!

RPigott commented on sway/tree/container.c:
```diff
@@ -1410,6 +1410,28 @@ void container_replace(struct sway_container *container,
+		list_t *siblings = container_get_siblings(child);
+		if (siblings->length == 1) {
+			enum sway_container_layout current = container_parent_layout(child);
+			if (container_is_floating(child)) {
```
If #5785 gets merged I won't need this.

RPigott commented:
Rebased on master again so the CI can build.

Xyene commented:
So far, so good. I haven't encountered any major issues in the last 2 days, so that's encouraging.

Here's a small one, though:

* start with `T[H[app* app]]`
* `focus parent`
* `splith` (multiple times)

The resulting layout is something like `T[H[H[H[H[...H[app app]]]]]`. In i3, the `splith`s remain a no-op.

@RedSoxFan, do you have any thoughts on this PR? I think it's roughly in a state where it could be merged: I think that some corner cases might still be wrong, but by and large it's less finicky than current master.

ChrisLane commented:
Just tested this branch, it seems to fix https://github.com/nwg-piotr/autotiling/issues/19

RPigott commented:
> The resulting layout is something like T[H[H[H[H[...H[app app]]]]]. In i3, the spliths remain a no-op.

Hmm I'm not able to reproduce. After the focus parent the H split is focused, right? In this branch and in i3 I get T[H[H[app app]*]] after the split and then all further splith's are no-ops. Actually it seems if you focus parent again in both this branch and i3 you can get more H[ by splith'ing the topmost H.

Something is fishy with workspace layouts still. Without setting `workspace_layout` in my config if I `layout tabbed` an empty workspace in i3 and open a new window I actually get `T[T[app]]` as if `workspace_layout tabbed` was set. I made several mistakes trying to create `T[H[app app]]` in i3 because of that.

Xyene commented:
> Hmm I'm not able to reproduce. After the focus parent the H split is focused, right? 

Yes, though oddly I haven't been able to reproduce this since. What I feel (since my memory isn't that good) was happening was the new container was getting focus, as opposed to the old one keeping it. At any rate, without being able to reproduce this, it's not worth worrying about now.

> Without setting workspace_layout in my config if I layout tabbed an empty workspace in i3 and open a new window I actually get T[T[app]] as if workspace_layout tabbed was set.

But... not quite? On i3 with `workspace_layout tabbed` and opening a single application, you get `H[T[app]]` (the workspace node layout is `splith`). But if you `layout tabbed` before opening the application (still `workspace_layout tabbed`), you get `T[T[app]]`. This matches the behavior in this PR, AFAICT.

Xyene commented:
I've been running this for the past week with no major issues. In general, I'd be comfortable with merging this before we get an authoritative response regarding `layout tabbed` on an empty workspace &mdash; that's a pretty niche scenario. Probably worth another round of rebasing to get this in a mergeable state.

On the "of some concern" list, I've noticed Firefox (84.0b3) geometries are wrong when floated out of the tiling layer. I'm not sure if this is caused by this patch (the geometry resetting?) or by a regression in Firefox, but at least I wasn't able to (easily) reproduce this issue in Sway master.

<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/1403503/99867290-5b29a500-2b86-11eb-8ee2-9453ff17eb51.png)

</details>

RPigott commented:
> On the "of some concern" list, I've noticed Firefox (84.0b3) geometries are wrong when floated out of the tiling layer.

I tried FF83 and 84 but it doesn't seem to happen for me. Is it reproducible everytime for you? Is there a chance it's fixed by #5812? I think I'm running that one too.

Xyene commented:
> Is it reproducible everytime for you? Is there a chance it's fixed by #5812? I think I'm running that one too.

Unfortunately yes, and partially.

#5812 does seem to prevent the issue from being permanent (once Firefox is in a bad state, it sends another configure a bit later, which after #5812 fixes the borders). Here's a [video](https://tbrindus.ca/files/dist/sway/5756-1.mp4).

There are at least two issues:

* When the window is floated, the border is drawn at the wrong position for 200ms (presumably it's waiting for Firefox to ack a configure?), and Firefox never does(?) so the transaction times out. With #5812, the original transaction still times out, but Firefox sends another one and "fixes" itself.

```
00:06:55.442 [sway/desktop/transaction.c:505] Transaction 0x5603c03bbae0 is ready
00:06:55.442 [sway/desktop/transaction.c:281] Applying transaction 0x5603c03bbae0
00:06:55.443 [sway/desktop/transaction.c:428] Transaction 0x5603c03d72b0 committing with 1 instructions
00:06:55.443 [sway/input/cursor.c:901] denying request to set cursor from unfocused client
00:06:55.443 [sway/input/cursor.c:901] denying request to set cursor from unfocused client
00:06:55.643 [sway/desktop/transaction.c:389] Transaction 0x5603c03d72b0 timed out (1 waiting)
00:06:55.643 [sway/desktop/transaction.c:281] Applying transaction 0x5603c03d72b0
```

* When the window is un-floated, there is a frame where the split indicator border is drawn incorrectly (my indicator color is red)

![image](https://user-images.githubusercontent.com/1403503/100560638-72524c00-3284-11eb-98f7-a6fe429fb5f6.png)

This is frame 118/550 of the video. I suspect this is happening as the floated container is being split on being added back into the tiling layer, prior to being squashed again.

RPigott commented:
Latest commit fixes #5783 but workspace {prev,next}_on_output will need the same treatment. Normally I'd make a different PR but it's just easier to have one big i3 compat branch for testing, especially when they might conflict. It should be easy enough to split later if we want.

As for firefox, it turns out #5812 was pretty buggy and I think it's related. I'm not sure yet how #5848 and #5857 will be resolved but I have a hunch their closure might address the firefox issue as well, or at least make it easier to debug.

Xyene reviewed (approved):
In the interest of getting more eyes on this and preventing this branch getting stale, I'll go ahead and merge this now. Irrespective of any outstanding bugs this is already a great improvement over the status quo. Thanks for working on this!
